src/boot/efi/random-seed.c

```diff
@@ -14,15 +14,28 @@
 
 #define EFI_RNG_GUID &(const EFI_GUID) EFI_RNG_PROTOCOL_GUID
 
+struct linux_efi_random_seed {
+      uint32_t size;
+      unsigned char seed[];
+};
+
+#define LINUX_EFI_RANDOM_SEED_TABLE_GUID \
+        &(EFI_GUID) { 0x1ce1e5bc, 0x7ceb, 0x42f2,  { 0x81, 0xe5, 0x8a, 0xad, 0xf1, 0x80, 0xf5, 0x7b } }
```

Comment:
should be `&(const EFI_GUID)`

Reply:
I wish, but InstallConfigurationTable wants non-const :(.

Reply:
Please no. This whole thing of forcing these guids into compound literals was a bad idea. It makes it impossible to directly initialize guid vars with these and marking them const even force tons of casts everywhere because most efi apis don't have the guids marked as consts. It's just inflexible and it's not like guids are expensive. It's literally 128 bits. Passing them around by copy internally everywhere is so much nicer (and the compiler knows what it's doing and can optimize where needed).

Reply:
Please no to what - original code, Lennart's comment, my follow up? What change should I make here?

Reply:
I'd really just cast the const away for InstallConfigurationTable. Sounds easiest and the most correct

Reply:
note that efi variables want pointers to GUIDs, not literal guids. so ther is no copy done either way. 

(in userspace, we pass around sd_id128 as true value, not via pointers)

Reply:
No to making this `&(const EFI_GUID)` or `&(EFI_GUID)`. I'd rather see the `&(EFI_GUID)` in the code where you need the guid or make a var on the stack and pass that around as needed.

With the forced cast you can't do this (which is the idiomatic way of making a EFI_GUID var):
```
EFI_GUID guid = LINUX_EFI_RANDOM_SEED_TABLE_GUID;
```

Reply:
In `missing_efi.h`, there's code like:

```c
#ifndef EFI_LOAD_FILE2_PROTOCOL_GUID                                                                          
#define EFI_LOAD_FILE2_PROTOCOL_GUID \                                                                        
        {0x4006c0c1, 0xfcb3, 0x403e, {0x99, 0x6d, 0x4a, 0x6c, 0x87, 0x24, 0xe0, 0x6d} }                       
#define EfiLoadFile2Protocol ((EFI_GUID)EFI_LOAD_FILE2_PROTOCOL_GUID)                                         
#endif   
```

I'll do the same thing here.

Reply:
Going to push this:

```diff
diff --git a/src/boot/efi/random-seed.c b/src/boot/efi/random-seed.c
index 0163e50647..20e4f00e81 100644
--- a/src/boot/efi/random-seed.c
+++ b/src/boot/efi/random-seed.c
@@ -20,7 +20,8 @@ struct linux_efi_random_seed {
 };
 
 #define LINUX_EFI_RANDOM_SEED_TABLE_GUID \
-        &(EFI_GUID) { 0x1ce1e5bc, 0x7ceb, 0x42f2,  { 0x81, 0xe5, 0x8a, 0xad, 0xf1, 0x80, 0xf5, 0x7b } }
+        { 0x1ce1e5bc, 0x7ceb, 0x42f2,  { 0x81, 0xe5, 0x8a, 0xad, 0xf1, 0x80, 0xf5, 0x7b } }
+#define LinuxEfiRandomSeedTable ((EFI_GUID)LINUX_EFI_RANDOM_SEED_TABLE_GUID)
 
 /* SHA256 gives us 256/8=32 bytes */
 #define HASH_VALUE_SIZE 32
@@ -146,8 +147,8 @@ EFI_STATUS process_random_seed(EFI_FILE *root_dir, RandomSeedMode mode) {
         sha256_process_bytes(HASH_LABEL, sizeof(HASH_LABEL) - 1, &hash);
 
         for (i = 0; i < ST->NumberOfTableEntries; ++i) {
-                if (!memcmp(LINUX_EFI_RANDOM_SEED_TABLE_GUID, &ST->ConfigurationTable[i].VendorGuid,
-                    sizeof(EFI_GUID))) {
+                if (!memcmp(&LinuxEfiRandomSeedTable, &ST->ConfigurationTable[i].VendorGuid,
+                            sizeof(EFI_GUID))) {
                         previous_seed_table = ST->ConfigurationTable[i].VendorTable;
                         break;
                 }
@@ -289,7 +290,7 @@ EFI_STATUS process_random_seed(EFI_FILE *root_dir, RandomSeedMode mode) {
         /* new_seed_table->seed = HASH(hash) */
         sha256_finish_ctx(&hash, new_seed_table->seed);
 
-        err = BS->InstallConfigurationTable(LINUX_EFI_RANDOM_SEED_TABLE_GUID, new_seed_table);
+        err = BS->InstallConfigurationTable(&LinuxEfiRandomSeedTable, new_seed_table);
         if (err != EFI_SUCCESS)
                 return log_error_status_stall(err, L"Failed to install EFI table for random seed: %r", err);
         TAKE_PTR(new_seed_table);
```

Reply:
The `EfiLoadFile2Protocol` emulates the static guids for gnu-efi compat. Those are supposed to go away, so please don't do that either.

Reply:
Oh, okay. So you want me to just declare a variable or something on the stack and use that?

Reply:
I'll do this:

```diff
diff --git a/src/boot/efi/random-seed.c b/src/boot/efi/random-seed.c
index 20e4f00e81..be37dae0e6 100644
--- a/src/boot/efi/random-seed.c
+++ b/src/boot/efi/random-seed.c
@@ -21,7 +21,6 @@ struct linux_efi_random_seed {
 
 #define LINUX_EFI_RANDOM_SEED_TABLE_GUID \
         { 0x1ce1e5bc, 0x7ceb, 0x42f2,  { 0x81, 0xe5, 0x8a, 0xad, 0xf1, 0x80, 0xf5, 0x7b } }
-#define LinuxEfiRandomSeedTable ((EFI_GUID)LINUX_EFI_RANDOM_SEED_TABLE_GUID)
 
 /* SHA256 gives us 256/8=32 bytes */
 #define HASH_VALUE_SIZE 32
@@ -120,6 +119,7 @@ static void validate_sha256(void) {
 
 EFI_STATUS process_random_seed(EFI_FILE *root_dir, RandomSeedMode mode) {
         _cleanup_erase_ uint8_t random_bytes[DESIRED_SEED_SIZE], hash_key[HASH_VALUE_SIZE];
+        EFI_GUID linux_efi_random_seed_table = LINUX_EFI_RANDOM_SEED_TABLE_GUID;
         _cleanup_free_ struct linux_efi_random_seed *new_seed_table = NULL;
         struct linux_efi_random_seed *previous_seed_table = NULL;
         _cleanup_free_ void *seed = NULL, *system_token = NULL;
@@ -147,7 +147,7 @@ EFI_STATUS process_random_seed(EFI_FILE *root_dir, RandomSeedMode mode) {
         sha256_process_bytes(HASH_LABEL, sizeof(HASH_LABEL) - 1, &hash);
 
         for (i = 0; i < ST->NumberOfTableEntries; ++i) {
-                if (!memcmp(&LinuxEfiRandomSeedTable, &ST->ConfigurationTable[i].VendorGuid,
+                if (!memcmp(&linux_efi_random_seed_table, &ST->ConfigurationTable[i].VendorGuid,
                             sizeof(EFI_GUID))) {
                         previous_seed_table = ST->ConfigurationTable[i].VendorTable;
                         break;
@@ -290,7 +290,7 @@ EFI_STATUS process_random_seed(EFI_FILE *root_dir, RandomSeedMode mode) {
         /* new_seed_table->seed = HASH(hash) */
         sha256_finish_ctx(&hash, new_seed_table->seed);
 
-        err = BS->InstallConfigurationTable(&LinuxEfiRandomSeedTable, new_seed_table);
+        err = BS->InstallConfigurationTable(&linux_efi_random_seed_table, new_seed_table);
         if (err != EFI_SUCCESS)
                 return log_error_status_stall(err, L"Failed to install EFI table for random seed: %r", err);
         TAKE_PTR(new_seed_table);
```

Which is what I think you want?

Reply:
Yes. Alternatively, use compound literals on the call site like this: `&(EFI_GUID)LINUX_EFI_RANDOM_SEED_TABLE_GUID`.

Reply:
Oh at the call site, better. Okay will do.

Reply:
Done.